src/cli/index.js

```diff
@@ -234,68 +235,129 @@ export function main({
-                resolve(runEventuallyWithNetwork(mutexPort));
-              });
-            }
-          })
-          .on('error', () => {
-            // No server to listen to ? Let's retry to become the next server then.
-            process.nextTick(() => {
-              resolve(runEventuallyWithNetwork(mutexPort));
-            });
+        // No socket must timeout
+        server.timeout = 0;
+
+        // If we fail to setup the server, we ask the existing one for its name
+        server.on('error', () => {
+          reportServerName();
+        });
+
+        // If we succeed, keep track of all the connected sockets to delete them later
+        server.on('connection', socket => {
+          clients.add(socket);
+          socket.on('close', () => {
+            clients.delete(socket);
           });
-      });
+        });
+
+        server.listen(connectionOptions, () => {
+          // Don't forget to kill the sockets if we're being killed via signals
+          onDeath(killSockets);
+
+          // Also kill the sockets if we finish, whether it's a success or a failure
+          run().then(
+            res => {
+              killSockets();
+              resolve(res);
+            },
+            err => {
+              killSockets();
+              reject(err);
+            },
+          );
+        });
+
+        function manager(request, response) {
+          response.writeHead(200);
+          response.write(config.cwd);
+          response.end();
+        }
 
-      const onServerEnd = async () => {
-        for (const client of clients) {
+        function killSockets() {
           try {
-            client.destroy();
+            server.close();
           } catch (err) {
-            // pass
+            // best effort
+          }
+
+          for (const socket of clients) {
+            try {
+              socket.destroy();
+            } catch (err) {
+              // best effort
+            }
           }
+
+          // If the process hasn't exited in the next 5s, it has stalled and we abort
+          const timeout = setTimeout(() => {
+            console.error('Process stalled');
+            if (process._getActiveHandles) {
+              console.error('Active handles:');
+              // $FlowFixMe: getActiveHandles is undocumented, but it exists
+              for (const handle of process._getActiveHandles()) {
+                console.error(`  - ${handle.constructor.name}`);
+              }
+            }
+            // eslint-disable-next-line no-process-exit
+            process.exit(1);
+          }, 5000);
+
+          // This timeout must not prevent us from exiting
+          // $FlowFixMe: Node's setTimeout returns a Timeout, not a Number
+          timeout.unref();
         }
+      }
 
-        await server.close();
-        server.unref();
-      };
+      function reportServerName() {
+        const request = http.get(connectionOptions, response => {
+          const buffers = [];
```

Comment:
I'd move this closer to where it is first used: right above `response.on('data'` to clarify the reason for it.